src/drivers/intel/cavs/ipc.c

```diff
@@ -133,12 +162,40 @@ static void ipc_irq_handler(void *arg)
 	}
 }
 
+#if CAVS_VERSION >= CAVS_VERSION_1_8
+static struct sof_ipc_cmd_hdr *ipc_cavs_read_msg(void)
+{
+	struct sof_ipc_cmd_hdr *hdr = _ipc->comp_data;
+	uint32_t dr;
+	uint32_t dd;
+	uint32_t *payload;
+
+	dr = ipc_read(IPC_DIPCTDR);
+	dd = ipc_read(IPC_DIPCTDD);
+
+	if (dr & IPC_DIPCTDR_MSG_REG) {
+		hdr->cmd = IPC_DIPCTDR_GLB_CMD(dr);
+		hdr->size = sizeof(*hdr) + IPC_DIPCTDD_PAYLOAD_SIZE(dd);
+		payload = (uint32_t *)(hdr + 1);
+		*payload = IPC_DIPCTDD_PAYLOAD(dd);
+		dcache_writeback_region(hdr, hdr->size);
+	} else {
+		hdr = mailbox_validate();
+	}
+
+	return hdr;
+}
+#endif
+
 void ipc_platform_do_cmd(struct ipc *ipc)
 {
 	struct sof_ipc_cmd_hdr *hdr;
 
+#if CAVS_VERSION >= CAVS_VERSION_1_8
+	hdr = ipc_cavs_read_msg();
+#else
 	hdr = mailbox_validate();
-
+#endif
```

Comment:
i would have ipc_cavs_read_msg() return NULL if IPC_DIPCTDR_MSG_REG is not set, and call mailbox_validate if (!hdr).